File: ios/Sqwack/Stores/SqwackStore.swift
import Foundation
import Observation

/// Root store. Holds one NodeConnection per registered daemon; every query
/// takes a machineId or `nil` (= all machines) so multi-machine needs no
/// schema change — only more nodes in the array and a picker in the UI.
@Observable
final class SqwackStore {
    private(set) var nodes: [NodeConnection] = []
    var selectedMachineId: String?

    private static let endpointsKey = "sqwack.endpoints"

    init() {
        for saved in UserDefaults.standard.stringArray(forKey: Self.endpointsKey) ?? [] {
            if let url = URL(string: saved) {
                nodes.append(NodeConnection(endpoint: url, credentialRef: saved))
            }
        }
        #if DEBUG
        // Dev/testing hook: pre-pair from the environment (used by UI automation).
        let env = ProcessInfo.processInfo.environment
        if nodes.isEmpty, let raw = env["SQWACK_ENDPOINT"], let url = URL(string: raw), let token = env["SQWACK_TOKEN"] {
            addNode(endpoint: url, token: token)
        }
        #endif
    }

    var isPaired: Bool { !nodes.isEmpty }

    func connectAll() {
        nodes.forEach { $0.connect() }
    }

    func addNode(endpoint: URL, token: String) {
        let ref = endpoint.absoluteString
        Keychain.save(token, ref: ref)
        var saved = UserDefaults.standard.stringArray(forKey: Self.endpointsKey) ?? []
        if !saved.contains(ref) { saved.append(ref) }
        UserDefaults.standard.set(saved, forKey: Self.endpointsKey)
        let node = NodeConnection(endpoint: endpoint, credentialRef: ref)
        nodes.append(node)
        node.connect()
    }

    func removeNode(_ node: NodeConnection) {
        if selectedMachineId == node.machine?.id { selectedMachineId = nil }
        node.disconnect()
        Keychain.delete(ref: node.credentialRef)
        var saved = UserDefaults.standard.stringArray(forKey: Self.endpointsKey) ?? []
        saved.removeAll { $0 == node.credentialRef }
        UserDefaults.standard.set(saved, forKey: Self.endpointsKey)
        nodes.removeAll { $0 === node }
    }

    func refreshUsage(provider: String? = nil, machineId: String? = nil) async {
        for node in nodes where machineId == nil || node.machine?.id == machineId {
            await node.refreshUsage(provider: provider)
        }
    }

    func refreshAll(machineId: String? = nil) async {
        for node in nodes where machineId == nil || node.machine?.id == machineId {
            await node.refreshSnapshot()
        }
    }

    func refreshAgents(machineId: String? = nil) async {
        await refreshAll(machineId: machineId)
    }

    func refreshProcessesOnly(machineId: String? = nil) async {
        for node in nodes where machineId == nil || node.machine?.id == machineId {
            await node.refreshProcesses()
        }
    }

    func refreshProcesses(machineId: String? = nil) async {
        await refreshAll(machineId: machineId)
    }

    var lastError: String? {
        nodes.compactMap(\.lastError).first
    }

    @MainActor
    func clearError() {
        nodes.forEach { $0.clearError() }
    }

    // MARK: - Aggregation (machineId == nil means "all machines")

    func sessions(machineId: String? = nil) -> [AgentSession] {
        nodes
            .flatMap { $0.sessions.values }
            .filter { machineId == nil || $0.machineId == machineId }
            .filter { !($0.source == "claude-process" && $0.state == .idle) }
            .sorted { $0.updatedAt > $1.updatedAt }
    }

    func processes(machineId: String? = nil) -> [DevProcess] {
        nodes
            .flatMap(\.processes)
            .filter { machineId == nil || $0.machineId == machineId }
            .sorted { ($0.port ?? 0) < ($1.port ?? 0) }
    }

    func machineName(for machineId: String) -> String {
        nodes.first { $0.machine?.id == machineId }?.machine?.name ?? machineId
    }

    /// MVP: system stats of the first (only) machine. Multi-machine: key by machineId.
    var system: SystemSnapshot? { nodes.first?.system }
    var topProcesses: [ProcessMetric] { nodes.first?.topProcesses ?? [] }
    var activity: [ActivityItem] {
        nodes.flatMap(\.activity).sorted { $0.timestamp > $1.timestamp }
    }
    func activity(machineId: String? = nil) -> [ActivityItem] {
        nodes
            .filter { machineId == nil || $0.machine?.id == machineId }
            .flatMap(\.activity)
            .sorted { $0.timestamp > $1.timestamp }
    }
    var machineName: String { nodes.first?.machine?.name ?? "" }
    var machineInfo: String {
        guard let m = nodes.first?.machine else { return "" }
        return "\(m.platform)/\(m.architecture)"
    }
    var daemonVersion: String { nodes.first?.machine?.daemonVersion ?? "0.1.0" }

    func usage(machineId: String? = nil) -> [ProviderUsage] {
        let values = nodes
            .filter { machineId == nil || $0.machine?.id == machineId }
            .flatMap(\.usage)
        return Dictionary(grouping: values, by: \.provider)
            .compactMap { $0.value.max { $0.collectedAt < $1.collectedAt } }
            .sorted { $0.provider < $1.provider }
    }
    var usage: [ProviderUsage] { usage() }

    var hermesNodes: [NodeConnection] {
        nodes.filter {
            (selectedMachineId == nil || $0.machine?.id == selectedMachineId)
                && !($0.hermes?.gateways.isEmpty ?? true)
        }
    }

    var hasHermes: Bool { !hermesNodes.isEmpty }

    /// Global status = worst status across machines (attention > failure > working > quiet).
    func status(machineId: String? = nil) -> SqwackStatus {
        nodes.filter { machineId == nil || $0.machine?.id == machineId }.map(\.status).max() ?? .quiet
    }
    var globalStatus: SqwackStatus { status() }

    var attention: [AgentSession] {
        sessions().filter { $0.state == .needsInput || $0.state == .failed }
    }

    /// Sessions worth showing on the ambient board: anything active, plus
    /// recently finished ones (done/failed fade out after an hour).
    func boardSessions(machineId: String? = nil) -> [AgentSession] {
        sessions(machineId: machineId).filter { session in
            switch session.state {
            case .working, .needsInput: true
            case .done, .failed: session.updatedAt > .now.addingTimeInterval(-3600)
            case .idle, .unknown: session.updatedAt > .now.addingTimeInterval(-900)
            }
        }
    }
    var boardSessions: [AgentSession] { boardSessions() }

    var anyConnected: Bool { nodes.contains { $0.connectionState == .connected } }
    var connectedMachineNames: [String] {
        nodes.filter { $0.connectionState == .connected }.compactMap { $0.machine?.name }
    }
}

File: android/app/src/main/java/com/sqwack/SqwackStore.kt
package com.sqwack

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import java.time.Instant

/// Root store. Holds one NodeConnection per registered daemon; every query
/// takes a machineId or `null` (= all machines) so multi-machine needs no
/// schema change — only more nodes in the list and a picker in the UI.
class SqwackStore(context: Context) {
    // ponytail: app-private SharedPreferences, not EncryptedSharedPreferences.
    // Jetpack Security Crypto is deprecated and the file is already inside the
    // app sandbox on a full-disk-encrypted device. Swap it if the tablet is shared.
    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences("sqwack", Context.MODE_PRIVATE)

    val nodes = mutableStateListOf<NodeConnection>()
    var selectedMachineId by mutableStateOf<String?>(null)

    init {
        prefs.getStringSet(ENDPOINTS_KEY, emptySet()).orEmpty().sorted().forEach { saved ->
            nodes.add(NodeConnection(saved, saved, prefs))
        }
    }

    val isPaired: Boolean get() = nodes.isNotEmpty()

    fun connectAll() = nodes.forEach { it.connect() }

    fun addNode(endpoint: String, token: String) {
        prefs.edit()
            .putString(NodeConnection.TOKEN_PREFIX + endpoint, token)
            .putStringSet(ENDPOINTS_KEY, savedEndpoints() + endpoint)
            .apply()
        val node = NodeConnection(endpoint, endpoint, prefs)
        nodes.add(node)
        node.connect()
    }

    fun removeNode(node: NodeConnection) {
        if (selectedMachineId == node.machine?.id) selectedMachineId = null
        node.disconnect()
        prefs.edit()
            .remove(NodeConnection.TOKEN_PREFIX + node.credentialRef)
            .putStringSet(ENDPOINTS_KEY, savedEndpoints() - node.credentialRef)
            .apply()
        nodes.remove(node)
    }

    private fun savedEndpoints(): Set<String> = prefs.getStringSet(ENDPOINTS_KEY, emptySet()).orEmpty().toSet()

    private fun matching(machineId: String?) = nodes.filter { machineId == null || it.machine?.id == machineId }

    suspend fun refreshUsage(provider: String? = null, machineId: String? = null) {
        matching(machineId).forEach { it.refreshUsage(provider) }
    }

    suspend fun refreshAll(machineId: String? = null) {
        matching(machineId).forEach { it.refreshSnapshot() }
    }

    suspend fun refreshAgents(machineId: String? = null) = refreshAll(machineId)

    suspend fun refreshProcessesOnly(machineId: String? = null) {
        matching(machineId).forEach { it.refreshProcesses() }
    }

    suspend fun refreshProcesses(machineId: String? = null) = refreshAll(machineId)

    val lastError: String? get() = nodes.firstNotNullOfOrNull { it.lastError }

    fun clearError() = nodes.forEach { it.clearError() }

    // MARK: - Aggregation (machineId == null means "all machines")

    fun sessions(machineId: String? = null): List<AgentSession> =
        nodes.flatMap { it.sessions.values }
            .filter { machineId == null || it.machineId == machineId }
            .filterNot { it.source == "claude-process" && it.state == AgentState.IDLE }
            .sortedByDescending { it.updatedAt }

    fun processes(machineId: String? = null): List<DevProcess> =
        nodes.flatMap { it.processes }
            .filter { machineId == null || it.machineId == machineId }
            .sortedBy { it.port ?: 0 }

    fun machineName(machineId: String): String =
        nodes.firstOrNull { it.machine?.id == machineId }?.machine?.name ?: machineId

    /// MVP: system stats of the first (only) machine. Multi-machine: key by machineId.
    val system: SystemSnapshot? get() = nodes.firstOrNull()?.system
    val topProcesses: List<ProcessMetric> get() = nodes.firstOrNull()?.topProcesses.orEmpty()

    fun activity(machineId: String? = null): List<ActivityItem> =
        matching(machineId).flatMap { it.activity }.sortedByDescending { it.timestamp }

    val machineName: String get() = nodes.firstOrNull()?.machine?.name.orEmpty()
    val daemonVersion: String get() = nodes.firstOrNull()?.machine?.daemonVersion ?: "0.1.0"

    fun usage(machineId: String? = null): List<ProviderUsage> =
        matching(machineId).flatMap { it.usage }
            .groupBy { it.provider }
            .mapNotNull { (_, items) -> items.maxByOrNull { it.collectedAt } }
            .sortedBy { it.provider }

    val hermesNodes: List<NodeConnection>
        get() = nodes.filter {
            (selectedMachineId == null || it.machine?.id == selectedMachineId) &&
                it.hermes?.gateways?.isNotEmpty() == true
        }

    val hasHermes: Boolean get() = hermesNodes.isNotEmpty()

    /// Global status = worst status across machines (attention > failure > working > quiet).
    fun status(machineId: String? = null): SqwackStatus =
        matching(machineId).maxByOrNull { it.status.rank }?.status ?: SqwackStatus.QUIET

    val attention: List<AgentSession>
        get() = sessions().filter { it.state == AgentState.NEEDS_INPUT || it.state == AgentState.FAILED }

    /// Sessions worth showing on the ambient board: anything active, plus
    /// recently finished ones (done/failed fade out after an hour).
    fun boardSessions(machineId: String? = null): List<AgentSession> {
        val now = Instant.now()
        return sessions(machineId).filter { session ->
            when (session.state) {
                AgentState.WORKING, AgentState.NEEDS_INPUT -> true
                AgentState.DONE, AgentState.FAILED -> session.updatedAt > now.minusSeconds(3600)
                AgentState.IDLE, AgentState.UNKNOWN -> session.updatedAt > now.minusSeconds(900)
            }
        }
    }

    val anyConnected: Boolean get() = nodes.any { it.connectionState == ConnectionState.CONNECTED }
    val connectedMachineNames: List<String>
        get() = nodes.filter { it.connectionState == ConnectionState.CONNECTED }.mapNotNull { it.machine?.name }

    private companion object {
        const val ENDPOINTS_KEY = "sqwack.endpoints"
    }
}
